File: framework/web/src/main/java/cn/bestwu/simpleframework/web/NavFilter.java
package cn.bestwu.simpleframework.web;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Optional;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.boot.actuate.autoconfigure.endpoint.web.WebEndpointProperties;
import org.springframework.core.io.Resource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.web.filter.OncePerRequestFilter;

/**
 * @author Peter Wu
 */
public class NavFilter extends OncePerRequestFilter {

  private static final String staticLocations = "classpath:/META-INF/resources";
  private final WebEndpointProperties webEndpointProperties;
  private final ResourceLoader resourceLoader;

  public NavFilter(
      WebEndpointProperties webEndpointProperties,
      ResourceLoader resourceLoader) {
    this.webEndpointProperties = webEndpointProperties;
    this.resourceLoader = resourceLoader;
  }


  @Override
  protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
      FilterChain filterChain) throws ServletException, IOException {
    String servletPath = request.getServletPath();
    if (servletPath.startsWith("/doc")) {
      response.sendRedirect(
          request.getContextPath() + webEndpointProperties.getBasePath() + "/doc" + servletPath
              .substring(4));
      return;
    }
    if (servletPath.startsWith("/logs")) {
      response.sendRedirect(
          request.getContextPath() + webEndpointProperties.getBasePath() + "/logs" + servletPath
              .substring(5));
      return;
    }
    if ("/actuator/doc".equals(servletPath)) {
      String name = staticLocations + webEndpointProperties.getBasePath() + "/doc/";
      Resource resource = resourceLoader.getResource(name);
      if (resource.exists()) {
        File dic = resource.getFile();
        if (dic.exists()) {
          File file = new File(dic, "v1.0.html");
          if (file.exists()) {
            response.sendRedirect(
                request.getContextPath() + webEndpointProperties.getBasePath() + "/doc/v1.0.html");
            return;
          } else {
            Optional<File> first = Arrays.stream(dic.listFiles())
                .filter(f -> f.isFile() && f.getName().endsWith(".html"))
                .min(Comparator.comparing(File::getName));
            if (first.isPresent()) {
              response.sendRedirect(
                  request.getContextPath() + webEndpointProperties.getBasePath() + "/doc/" + first
                      .get()
                      .getName());
              return;
            }
          }
        }
      }
    }
    filterChain.doFilter(request, response);
  }
}

File: web/src/main/kotlin/top/bettercode/summer/config/NavFilter.kt
package top.bettercode.summer.config

import org.springframework.boot.actuate.autoconfigure.endpoint.web.WebEndpointProperties
import org.springframework.core.Ordered
import org.springframework.core.io.ResourceLoader
import org.springframework.core.io.support.PathMatchingResourcePatternResolver
import org.springframework.web.filter.OncePerRequestFilter
import java.io.File
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


/**
 * @author Peter Wu
 */
class NavFilter(
        private val webEndpointProperties: WebEndpointProperties,
        private val resourceLoader: ResourceLoader
) : OncePerRequestFilter(), Ordered {
    var resolver = PathMatchingResourcePatternResolver()

    override fun doFilterInternal(
            request: HttpServletRequest,
            response: HttpServletResponse,
            filterChain: FilterChain
    ) {
        val servletPath = request.servletPath
        if (servletPath.startsWith("/doc")) {
            response.sendRedirect(
                    request.contextPath + webEndpointProperties.basePath + "/doc" + servletPath
                            .substring(4)
            )
            return
        }
        if (servletPath.startsWith("/logs")) {
            response.sendRedirect(
                    request.contextPath + webEndpointProperties.basePath + "/logs" + servletPath
                            .substring(5)
            )
            return
        }
        if (webEndpointProperties.basePath + "/doc" == servletPath || webEndpointProperties.basePath + "/doc/" == servletPath) {
            val name = STATIC_LOCATIONS + webEndpointProperties.basePath + "/doc/"
            val resource = resourceLoader.getResource(name)
            if (resource.exists()) {
                val url = resource.url.toString()
                if (url.startsWith("jar:")) {
                    val resources = resolver.getResources("$name**").map { it.url.path.substringAfter(resource.url.path) }.filter { it.matches("^v.*\\.html$".toRegex()) }
                    val last = resources
                            .sortedWith { o1, o2 -> o1.compareTo(o2) }.lastOrNull()
                    if (last != null) {
                        response.sendRedirect(
                                request.contextPath + webEndpointProperties.basePath + "/doc/" + last
                        )
                        return
                    }
                } else if (resource.file.exists()) {
                    val last = Arrays.stream(resource.file.listFiles())
                            .filter { f: File -> f.isFile && f.name.matches("^v.*\\.html$".toRegex()) }
                            .max { o1, o2 -> o1.name.compareTo(o2.name) }
                    if (last.isPresent) {
                        response.sendRedirect(
                                request.contextPath + webEndpointProperties.basePath + "/doc/" + last
                                        .get()
                                        .name
                        )
                        return
                    }
                }
            }
        }
        filterChain.doFilter(request, response)
    }

    companion object {
        private const val STATIC_LOCATIONS = "classpath:/META-INF/resources"
    }

    override fun getOrder(): Int {
        return Ordered.HIGHEST_PRECEDENCE + 1
    }
}